src/core/environment.go

```diff
@@ -307,18 +334,25 @@ func (env *EnvironmentType) IsContainer() bool {
 		k8sServiceAcct = "/var/run/secrets/kubernetes.io/serviceaccount"
 	)
 
-	for _, filename := range []string{dockerEnv, containerEnv, k8sServiceAcct} {
-		if _, err := os.Stat(filename); err == nil {
-			log.Debugf("is a container because (%s) exists", filename)
-			return true
+	res, err, _ := singleflightGroup.Do(IsContainerKey, func() (interface{}, error) {
+		for _, filename := range []string{dockerEnv, containerEnv, k8sServiceAcct} {
+			if _, err := os.Stat(filename); err == nil {
+				log.Debugf("is a container because (%s) exists", filename)
+				return true, nil
+			}
 		}
-	}
-	// v1 check
-	if result, err := cGroupV1Check(selfCgroup); err == nil && result {
-		return result
+		// v1 check
+		if result, err := cGroupV1Check(selfCgroup); err == nil && result {
+			return result, err
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		log.Warnf("unable to retrieve values from cache (%v)", err)
```

Comment:
(Optionally follow logging convention)
```suggestion
		log.Warnf("Unable to retrieve values from cache (%v)", err)
```